week4/solutions/scrabble/tests/test_hand.py

```diff
@@ -0,0 +1,480 @@
+    mock_get_words_from_user.side_effect = [None, None, None]
+    mock_display_total_score.side_effect = [None, None, None]
+    hand.play()
+
+    self.assertTrue(mock_get_words_from_user.called)
+    self.assertTrue(mock_display_total_score.called)
+
+  def test(self):
+    self.mock_set_word_function()
+
+
+# Hand.update()
+class HandUpdateTest1(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+    hand.updated_hand = {'e':5, 'v':2, 'n':4, 'i':2, 'l':2}
+    hand.update('nvn')
+
+    self.assertEqual(hand.updated_hand, {'e':5, 'v':1, 'n':2, 'i':2, 'l':2})
+
+class HandUpdateTest2(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+    hand.updated_hand = {'x':5, 'v':2, 'n':4, 'i':2, 'l':2}
+    hand.update('xnxnvvxxl')
+
+    self.assertEqual(hand.updated_hand, {'x':1, 'v':0, 'n':2, 'i':2, 'l':1})
+
+class HandUpdateTest3(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+    hand.updated_hand = {'x':0, 'v':2, 'n':4, 'i':0, 'l':2}
+    hand.update('vnlv')
+
+    self.assertEqual(hand.updated_hand, {'x':0, 'v':0, 'n':3, 'i':0, 'l':1})
+
+  
+# Hand.check_user_is_playing()
+class HandCheckUserIsPlayingTest1(unittest.TestCase):
+  def test(self):
+    hand = Hand()
+
+    self.assertEqual(hand.check_user_is_playing(Word('.')), False)
+
+class HandCheckUserIsPlayingTest2(unittest.TestCase):
+  @unittest.mock.patch('hand.Hand.evaluate_word')
+  def mock_check_user_is_playing_function(self, mock_evaluate_word):
+    hand = Hand()
+    hand.set({'e':1, 'v':1, 'n':2, 'i':2, 'l':1})
+    hand.updated_hand = {'e':1, 'v':1, 'n':2, 'i':2, 'l':1}
+    hand.check_user_is_playing(Word('evil'))
+
+    self.assertTrue(mock_evaluate_word.called)
+
+  def test(self):
+    self.mock_check_user_is_playing_function()
+
+class HandCheckUserIsPlayingTest3(unittest.TestCase):
+  @unittest.mock.patch('hand.Hand.evaluate_word')
+  def mock_check_user_is_playing_function(self, mock_evaluate_word):
+    hand = Hand()
+    hand.set({'e':1, 'v':1, 'n':2, 'i':2, 'l':1})
+    hand.updated_hand = {'e':1, 'v':1, 'n':2, 'i':2, 'l':1}
+    hand.check_user_is_playing(Word('soup'))
+
+    self.assertTrue(mock_evaluate_word.called)
+
+  def test(self):
+    self.mock_check_user_is_playing_function()
+
+
+# Hand.evaluate_word()
+class HandEvaluateWordTest1(unittest.TestCase):
+  @unittest.mock.patch('score.Score.calculate')
+  @unittest.mock.patch('hand.Hand.update')
+  def mock_evaluate_word_function(self, mock_calculate_score, mock_update_hand):
+    hand = Hand()
+    hand.set({'e':1, 'v':1, 'n':2, 'i':2, 'l':1})
+    hand.updated_hand = {'e':1, 'v':1, 'n':2, 'i':2, 'l':1}
+    hand.evaluate_word(Word('evil'))
+    
+    self.assertTrue(mock_calculate_score.called)
+    self.assertTrue(mock_update_hand.called)
+
+  def test(self):
+    self.mock_evaluate_word_function()
+
+class HandEvaluateWordTest2(unittest.TestCase):
+  @unittest.mock.patch('sys.stdout', new_callable=io.StringIO)
+  def assert_stdout(self, expected_output, mock_stdout):
```

Comment:
Not sure if this is a good way to check print statements.

Reply:
First, notice how you call `hand = Hand()` over and over? If you put that into a constructor in the base class you can get that for free. Python docs has info on how to do that.

Second, unit tests should exercise _your_ logic. You don't test that stdout and printing actually work. We rely on some modules like those, trusting that they work. All you need to test is that the strings (output) generated by your code is exactly what should be printed.

Reply:
So, this isn't necessary:
```python
  def test(self):
    self.assert_stdout('invalid word ')
```